### PR TITLE
[WIP] Add support to download context file from Azure Blob Storage

### PR DESCRIPTION
**Pull request by yangtaokm**
Add support for downloading Build Contexts from Azure blob storage 

**Description**

- Add support for downloading Build Contexts from Azure blob storage with full blob storage url https://myaccount.blob.core.windows.net/container/path/to/context.tar.gz

- Forward to azureblob.go to hand the context downloading if the --context prefixed with https:// and the url matches AzureBlobStorageHostRegEx 

- use Azure Storage Blob SDK for Go https://github.com/azure/azure-storage-blob-go

- You will need to pass Azure Storage Account Access Key as an evironment variable named AZURE_STORAGE_ACCESS_KEY through Kubernetes Secrets

**Submitter Checklist**

These are the criteria that every PR should meet, please check them off as you
review them:

- [x] Includes [unit tests](../DEVELOPMENT.md#creating-a-pr)
- [please advise ] Adds integration tests if needed.

_See [the contribution guide](../CONTRIBUTING.md) for more details._

**Reviewer Notes**

- [ ] The code flow looks good. 
- [ ] Unit tests and or integration tests added.

**Release Notes**

Describe any changes here so maintainer can include it in the release notes, or delete this block.

```
Add support for downloading Build Contexts from Azure blob storage

```

### CODE REVIEW

**Comment by tejal29** [comment]
@yangtaokm please sign the Google CLA.

File: examples/pod-blobstroage.yaml [hunk 1]
```diff
@@ -0,0 +1,23 @@
...
+metadata:
+  name: kaniko
+spec:
+  containers:
+  - name: kaniko
+    image: gcr.io/kaniko-project/executor:latest
+    args: ["--dockerfile=<path to Dockerfile within the build context>",
+            "--context=https://myaccount.blob.core.windows.net/container/path/to/context.tar.gz",
+            "--destination=<registry for image push>"]
+...
```
Comment (tejal29) [inline examples/pod-blobstroage.yaml]: `examples` should point to a working yaml. Ideally users can just copy and replace the variables. Can you please remove the `...`

File: pkg/buildcontext/azureblob.go [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+
+// AzureBlob struct for Azure Blob Storage processing
+type AzureBlob struct {
+	context string
+}
+
+// Download context file from given azure blob storage url and unpack it to BuildContextDir
+func (b *AzureBlob) UnpackTarFromBuildContext() (string, error) {
+
+	//Get Azure_STORAGE_ACCESS_KEY from environment variables
```
Comment (tejal29) [inline pkg/buildcontext/azureblob.go]:
```suggestion
	// Get Azure_STORAGE_ACCESS_KEY from environment variables
```

File: pkg/buildcontext/azureblob.go [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+	accountKey := os.Getenv("AZURE_STORAGE_ACCESS_KEY")
+	if len(accountKey) == 0 {
+		return "", errors.New("AZURE_STORAGE_ACCESS_KEY environment variable is not set")
+	}
+
+	//Get storage accoutname for Azure Blob Storage
```
Comment (tejal29) [inline pkg/buildcontext/azureblob.go]:
```suggestion
	// Get storage accoutname for Azure Blob Storage
```

File: pkg/buildcontext/azureblob.go [hunk 4]
```diff
@@ -0,0 +1,82 @@
...
+	u, _ := url.Parse(b.context)
+	parts := azblob.NewBlobURLParts(*u)
+	accountName := strings.Split(parts.Host, ".")[0]
+
+	//Generate credentail with accountname and accountkey
```
Comment (tejal29) [inline pkg/buildcontext/azureblob.go]:
```suggestion
	// Generate credentail with accountname and accountkey
```

File: pkg/constants/constants.go [hunk 5]
```diff
@@ -78,3 +79,10 @@ const (
 
 // ScratchEnvVars are the default environment variables needed for a scratch image.
 var ScratchEnvVars = []string{"PATH=/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin"}
+
+//AzureBlobStorageHostRegEx is ReqEX for Valid azure blob storage host suffix in url for AzureCloud, AzureChinaCloud, AzureGermanCloud and AzureUSGovernment
```
Comment (tejal29) [inline pkg/constants/constants.go]:
```suggestion
// AzureBlobStorageHostRegEx is ReqEX for Valid azure blob storage host suffix in url for AzureCloud, AzureChinaCloud, AzureGermanCloud and AzureUSGovernment
```

File: pkg/util/azureblob_util.go [hunk 6]
```diff
@@ -0,0 +1,36 @@
...
+
+package util
+
+import (
+	"regexp"
+
+	"github.com/GoogleContainerTools/kaniko/pkg/constants"
+)
+
+//Validate if the host url provided is with correct suffix for AzureCloud, AzureChinaCloud, AzureGermanCloud and AzureUSGovernment
```
Comment (tejal29) [inline pkg/util/azureblob_util.go]:
```suggestion
// Validate if the host url provided is with correct suffix for AzureCloud, AzureChinaCloud, AzureGermanCloud and AzureUSGovernment
```

File: pkg/util/azureblob_util.go [hunk 7]
```diff
@@ -0,0 +1,36 @@
...
+//RegEX for supported suffix defined in constants.AzureBlobStorageHostRegEx
```
Comment (tejal29) [inline pkg/util/azureblob_util.go]:
```suggestion
// RegEX for supported suffix defined in constants.AzureBlobStorageHostRegEx
```

File: pkg/buildcontext/azureblob.go [hunk 8]
```diff
@@ -0,0 +1,82 @@
...
+		return parts.Host, err
+	}
+
+	//Create directory and file for downloading context file
+	directory := constants.BuildContextDir
+	tarPath := filepath.Join(directory, constants.ContextTar)
+	if err := os.MkdirAll(directory, 0750); err != nil {
+		return directory, err
+	}
+	file, err := os.Create(tarPath)
```
Comment (tejal29) [inline pkg/buildcontext/azureblob.go]: i think you can re-use util.CreateFile https://github.com/GoogleContainerTools/kaniko/blob/master/pkg/buildcontext/gcs.go#L71
```
if err := util.CreateFile(tarPath, reader, 0600, 0, 0); err != nil {
		return "", err
	}
```

Comment (tejal29) [reply pkg/buildcontext/azureblob.go]: Also, looks like this piece of code is repeated even in [s3.go](https://github.com/GoogleContainerTools/kaniko/blob/8acab90decb3e75fe61de8dea0cdb8bd8c11cc78/pkg/buildcontext/s3.go#L59)
```
	directory := constants.BuildContextDir
	tarPath := filepath.Join(directory, constants.ContextTar)
	if err := os.MkdirAll(directory, 0750); err != nil {
		return directory, err
	}
	file, err := os.Create(tarPath)
	if err != nil {
		return directory, err
	}
```

Would you mind creating a util like
`util.setUpBuildContextAndTarpath`

Comment (yangtaokm) [reply pkg/buildcontext/azureblob.go]: Thanks @tejal29  for review and suggestion. I will make this improvement shortly. thanks!

Comment (tejal29) [reply pkg/buildcontext/azureblob.go]: thank you for your contribution! Looking forward to get this in!!

Comment (yangtaokm) [reply pkg/buildcontext/azureblob.go]: @tejal29 I have pushed some changes to reflect our discussion 

> i think you can re-use util.CreateFile https://github.com/GoogleContainerTools/kaniko/blob/master/pkg/buildcontext/gcs.go#L71
> 
> ```
> if err := util.CreateFile(tarPath, reader, 0600, 0, 0); err != nil {
> 		return "", err
> 	}
> ```
The CreateFile in util do not only create the file but also copy the content to the file through a reader. I intended to let the sdk to handle the file directly. I wrapped the make directory and create file part into a new function CreateTargetTarfile and reused in azureblob.go. Let me know how it looks for you.

Thanks!

**Comment by tejal29** [comment]
@yangtaokm Can you please sign the Google CLA.

**Comment by yangtaokm** [comment]
@googlebot I fixed it

**Comment by yangtaokm** [comment]
@tejal29 thank you again for the review, suggestion and accepting the PR. We definitely will use Kaniko with Azure Blob Storage going forward. Thanks!

**Pull request merged**